File: cpp/n559_NTreeDepth.cpp
// https://leetcode-cn.com/problems/maximum-depth-of-n-ary-tree/
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <queue>
using namespace std;
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    int maxDepth(Node* root) {
        if (root == NULL) return 0;
        int depth = 0;
        queue<vector<Node*>> my_queue;
        vector<Node*> tmp;
        tmp.push_back(root);
        my_queue.push(tmp);
        while (my_queue.front().size() != 0){
            tmp.clear();
            for (auto node: my_queue.front()){
                if (node == NULL || (node -> children).size() == 0) continue;
                for (auto child: node -> children) tmp.push_back(child);
            }
            my_queue.push(tmp);
            my_queue.pop();
            depth += 1;
        }
        return depth;
    }
};

int main(){
    return 0;
}

File: python/n559_NTreeDepth.py
# https://leetcode-cn.com/problems/maximum-depth-of-n-ary-tree/
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

class Solution:
    def __init__(self):
        self.depth = 0
        self.queue = []
        
    def maxDepth(self, root: 'Node') -> int:
        if root is None: return 0
        tmp_queue = []
        tmp_queue.append(root)
        self.queue.append(tmp_queue)
        while len(self.queue[0]) != 0:
            tmp_queue = []
            for node in self.queue[0]:
                if node is None or len(node.children) == 0:
                    continue
                for i in node.children:
                    tmp_queue.append(i)
            self.queue.append(tmp_queue)
            self.queue.pop(0)
            self.depth += 1
        return self.depth
        
    def maxDepth_a(self, root: 'Node'):
        if root is None: return 0
        if root.children == []: return 1
        height = [self.maxDepth_a(node) for node in root.children]
        return max(height) + 1
        

if __name__ == "__main__":
    pass
            